transport/server.cc

```diff
@@ -236,23 +239,44 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
             auto conn = make_shared<connection>(*this, server_addr, std::move(fd), std::move(addr));
             ++_connects;
             ++_connections;
-            // Move connection to the background, monitor for lifetime and errors.
-            (void)conn->process().then_wrapped([this, conn] (future<> f) {
-                --_connections;
-                try {
-                    f.get();
-                } catch (...) {
-                    clogger.debug("connection error: {}", std::current_exception());
-                }
+            return advertise_new_connection(conn).then([this, conn] {
+                // Move connection to the background, monitor for lifetime and errors.
+                (void)conn->process().then_wrapped([this, conn] (future<> f) {
+                    try {
+                        f.get();
+                    } catch (...) {
+                        clogger.debug("connection error: {}", std::current_exception());
+                    }
+                }).finally([this, conn] {
+                    --_connections;
+                    return unadvertise_connection(conn);
+                });
+                return stop_iteration::no;
             });
-            return stop_iteration::no;
         }).handle_exception([] (auto ep) {
             clogger.debug("accept failed: {}", ep);
-            return stop_iteration::no;
+            return make_ready_future<stop_iteration>(stop_iteration::no);
         });
     });
 }
 
+future<>
+cql_server::advertise_new_connection(shared_ptr<connection> conn) {
+    db::system_keyspace::client_data cd;
```

Comment:
This is introducing a dependency between system_keyspace and transport. The way to break the dependency is to create an interface/implementation pair and have transport and system_keyspace only talk to the interface, not each other. main.cc can mediate between them.

Reply:
> The problems you listed don't worry me
Good to hear that, but please note that now `Scylla version ... initialization completed.` can no longer be the last line in startup logs (because of compaction logs covering it). It doesn't hurt, but also doesn't look as "inviting". Don't you think it should be dealt with?

`[shard 0] init - Scylla version 666.development-0.20191029.767f636eb initialization completed.
[shard 0] compaction - Compacted 2 sstables to [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-102-big-Data.db:level=0, ]. 82010 bytes to 76766 (~93% of original) in 294ms = 0.25MB/s. ~256 total partitions merged to 2.
[shard 0] compaction - Compacting [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-99-big-Data.db:level=0, tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-102-big-Data.db:level=0, ]
[shard 0] compaction - Compacted 2 sstables to [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-105-big-Data.db:level=0, ]. 82002 bytes to 76768 (~93% of original) in 290ms = 0.25MB/s. ~256 total partitions merged to 2.
`

Reply:
I decoupled them, but in a simpler way. The code that announces connection/disconnection was stateless, so it was enough just to shuffle things around.